### ISSUE TITLE
Client prefetching

### ISSUE THREAD

**Issue by HereThereBeDragons** [cvmfs/cvmfs#3078]
@chrisburr from LHCb reported slow access times when super many small files [ O(1k) ] are being accessed from cvmfs (e.g. root/python)

Tracing performed on the containers provided showed the following behavior:
- Runtime
    - Highly correlated to number of `open()` commands (this does not mean unique files!)
    - Correlated with Readlink
    - Correlated to ReadExtendedAttr only for warm cache
    - Correlated to LookupPath only for warm cache
    
Cold cache performance:
- `snakemake --help` had 633 `open()` requests while loading 31 MB with a local cache hit rate of ~9% and taking 16 sec (avg 2 MB/s throughput)
    - 607 unique file requests
    - speed up: warm vs cold cache = 7.3x
    - speed up: hot vs warm cache = 2.3x
- `root -b -l -q -x /opt/micromamba/envs/test-env/tutorials/hsimple.C`  had 8016 `open()` requests while loading 400 MB with a local cache hit rate of nearly 85% and taking 53 sec (avg 20 MB/s throughput)
    - 1142 unique file requests
    - speed up: warm vs cold cache = 10.1x
    - speed up: hot vs warm cache = 3.0x
( time is the runtime of the program including the cvmfs loading)
    
These results suggest good opportunities of improvement if some form of prefetching could be utilized.

**Comment by HereThereBeDragons** [comment]
Possible ideas so far:
- Preloading blobs/tarballs/bundles containing those many small files
- Have some `.cvmfsprefetch` that suggests which file should be prefetched together

**Comment by chrisburr** [comment]
A couple of other ideas to add from our discussions:

* The bundles could potentially improve warm cache performance as well by preloading the kernel with the files that you expect to be accessed next
* Some files can be automatically pre-fetched based upon the initial content. For example shared libraries will likely be read in their entirety. This can be done by inspecting the first few bytes of the file.
* For defining which files to pre-fetch we could start with some simple patterns (`lib/python3.9/site-packages/{pkg1,pkg2,pkg3}/__init__.py` triggers the fetching of the entire `{pkg1,pkg2,pkg3}` directory)
* The bundles should have a way of ignoring files that are larger than a specific size to prevent duplication
* Having bundles which trigger the loading of other bundles might be useful to avoid too much duplication of data.
* More advanced bundling can be done on the release manager by giving it a command to run while tracing. The file which triggers the bundle to be loaded can be given, for example:

```bash
/cvmfs/lhcb.cern.ch/lib/var/lib/LbEnv/2541/stable/linux-64/bin/lb-conda default/2022-09-20_13-08 python -c "import ROOT"
# Bundle starts with: /cvmfs/lhcbdev.cern.ch/conda/envs/default/2022-09-20_13-08/linux-64/lib/python3.9/site-packages/ROOT/__init__.py
# (ignore the fact this command is split across 2 repositories)
# Contains ~1400 files
# Maybe split into a few smaller bundles triggered by libCling.so etc.
```

```bash
/cvmfs/lhcb.cern.ch/lib/var/lib/LbEnv/2541/stable/linux-64/bin/python -c "pass"
# Bundle starts with: /cvmfs/lhcb.cern.ch/lib/var/lib/LbEnv/2541/stable/linux-64/bin/python3.9
# Contains ~48 files
```

```bash
/cvmfs/lhcb.cern.ch/lib/var/lib/LbEnv/2541/stable/linux-64/bin/lb-run --help
# Bundle starts with: /cvmfs/lhcb.cern.ch/lib/var/lib/LbEnv/2541/stable/linux-64/bin/lb-run
# Contains ~100 files but ignores the bin/python3.9 bundle (or triggers it directly)
```

```bash
/cvmfs/lhcb.cern.ch/lib/var/lib/LbEnv/2541/stable/linux-64/bin/lb-dev --help
# Bundle starts with: /cvmfs/lhcb.cern.ch/lib/var/lib/LbEnv/2541/stable/linux-64/bin/lb-run
# Contains ~100 files but ignores the bin/python3.9 bundle (or triggers it directly)
```

```bash
/cvmfs/lhcb.cern.ch/lib/var/lib/LbEnv/2541/stable/linux-64/bin/python -c 'import matplotlib'
# Bundle starts with: /cvmfs/lhcbdev.cern.ch/conda/envs/default/2022-09-20_13-08/linux-64/lib/python3.9/site-packages/matplotlib/__init__.py
# Contains ~46 files
# Also triggers loading bundles for numpy/__init__.py (181 files) PIL/__init__.py (46 files) (and maybe others)
# Uses the bin/python3.9 bundle before matplotlib/__init__.py is touched
```

**Comment by HereThereBeDragons** [comment]
Have a look at related/previous work: https://sft.its.cern.ch/jira/browse/CVM-1886 and its PRs #2741 #2742 #2743

**Comment by PerilousApricot** [comment]
Hi @chrisburr - I'm curious if you have a size distribution of the files opened via open()

**Comment by chrisburr** [comment]
I released some of the context to this issue was missing see slide 14-17 of https://indico.cern.ch/event/1079490/contributions/4939521/ and https://gitlab.cern.ch/cburr/cvmfs-client-performance-studies

I also have Laura's notes from the checks she performed in follow up to my initial report (which then led to this issue).

**Comment by DrDaveD** [comment]
The suggestion to prefetch all the chunks of shared libraries raises a red flag for me on this concept from the point of view of the distribution infrastructure.  In fact the primary reason cvmfs chunks larger files is to intentionally slow down clients to give other clients a chance to have their fair share of the server resources; if lots of clients suddenly request large files at once, it can be too much for the proxy cache servers to handle.  If prefetching is implemented my worry is that large batches of jobs requesting prefetching would overload the proxy caches.  If it were limited to just interactive sessions that would probably be OK.

**Comment by vvolkl** [comment]
Yes, that's indeed a concern, but we always rely on responsible usage of the infrastructure. The intention is certainly to make this feature opt-in and have it be enabled only where it makes sense, so lxplus but not lxbatch.

**Comment by vvolkl** [comment]
Implemented in #4002
